File: congreso_back_end/src/main/java/py/com/flextech/model/sistema/Usuario.java
package py.com.flextech.model.sistema;

import java.io.Serializable;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.Collection;
import java.util.List;

import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;

import com.fasterxml.jackson.annotation.JsonIgnore;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import py.com.flextech.model.checkin.Checkin;

@Entity
@Table(name = "SYS_USUARIO")
@Data
@EqualsAndHashCode(callSuper = false)
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Usuario  implements Serializable , UserDetails  {
	/**
	 * 
	 */
	private static final long serialVersionUID = -4854234903205150197L;

	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	@Column(name = "ID_USUARIO")
	private Long id;
	
	@NotNull
	@Column(name = "FC_REGISTRO")
	private LocalDateTime fechaRegistro;

	@NotNull
	@Column(name = "NOMBRE_COMPLETO")
	private String nombreCompleto;

	@NotNull
	@Column(name = "EMAIL", unique = true)
	private String email;

	@NotNull
	@Column(name = "SENHA")
	private String senha;

	@NotNull
	@Column(name = "TELEFONO")
	private String telefono;

	@NotNull
	@Column(name = "INSTITUCION")
	private String institucion;

	@NotNull
	@Column(name = "REGISTRO_ACADEMICO")
	private String registroAcademico;

	@NotNull
	@Column(name = "SEMESTRE")
	private String semestre;

	@NotNull
	@Column(name = "SECCION")
	private String seccion;

	@NotNull
	@Column(name = "PAIS")
	private String pais;
	
	//ACTIVACION
	@Column(name = "FC_ACTIVACION")
	private LocalDateTime fechaActivacion;

	@NotNull
	@Column(name = "BO_ACTIVADO")
	private Boolean isActivado;
	
	//PAGO
	@NotNull
	@Column(name = "BO_IS_PAGO")
	private Boolean isPago;

	@Column(name = "FC_PAGO")
	private LocalDateTime fechaPago;

	@Column(name = "NR_COMPROBANTE")
	private String nrComprobante;
	
	@Column(name = "USUARIO_PAGO")
	private String usuarioPago;

	@Column(name = "VL_PAGO")
	private BigDecimal montoPago;

	@Column(name = "BO_EXONERADO")
	private Boolean isExonerado;
	
	@Column(name = "OBS_ANULACION_PAGO")
	private String obsAnulacionPago;

	@NotNull
	@Column(name = "UUID")
	private String uuid;
	
	@Column(name = "UUID_BIN", insertable = false, updatable = false)
	private byte[] uuidBin;

	@Column(name = "BO_ADMIN")
	private Boolean isAdmin;

	@Column(name = "BO_FINANCIERO")
	private Boolean isFinanciero;
	
	@Column(name = "BO_STAFF")
	private Boolean isStaff;

	@Column(name = "BO_CONGRESISTA")
	private Boolean isCongresista;
	
	@Column(name = "BO_INVITADO")
	private Boolean isInvitado;

	@Column(name = "BO_DISERTANTE")
	private Boolean isDisertante;
	
	@Column(name = "BO_AUDIOVISUAL")
	private Boolean isAudioVisual;

	@Column(name = "BO_IS_CHECK_IN")
	private Boolean isCheckIn;
	
	@OneToMany(mappedBy = "usuario", cascade = CascadeType.ALL)
	private List<Checkin> checkin;

	
	@JsonIgnore
	public Usuario(Long id) {
		this.id = id;
	}



	@JsonIgnore
	@Override
	public String getPassword() {
		return senha;
	}

	@JsonIgnore
	@Override
	public String getUsername() {
		return email;
	}

	@JsonIgnore
	@Override
	public boolean isAccountNonExpired() {
		return true;
	}

	@JsonIgnore
	@Override
	public boolean isAccountNonLocked() {
		return true;
	}

	@JsonIgnore
	@Override
	public boolean isCredentialsNonExpired() {
		return true;
	}

	@JsonIgnore
	@Override
	public boolean isEnabled() {
		return true;
	}

	@Override
	public Collection<? extends GrantedAuthority> getAuthorities() {
		// TODO Auto-generated method stub
		return null;
	}


}

File: congreso_frond_end/lib/modules/auth/models/usuario.dart
import 'package:congreso_evento/core/date_time_converter.dart';
import 'package:congreso_evento/core/i_simple_list_tile.dart';
import 'package:json_annotation/json_annotation.dart';

part 'usuario.g.dart';

@DateTimeConverter()
@JsonSerializable()
class Usuario implements ISimpleListTile {
  int? id;
  DateTime? fechaRegistro;
  String? nombreCompleto;
  String? email;
  String? senha;
  String? telefono;
  String? institucion;
  String? registroAcademico;
  String? semestre;
  String? seccion;
  String? pais;

  //Pago
  bool isPago;
  DateTime? fechaPago;
  double? montoPago;
  String? usuarioPago;
  bool? isExonerado;

  String? uuid;

  bool? recordarPassword = false;

  //ROLES
  bool? isAdmin = false;
  bool? isFinanciero = false;
  bool? isStaff = false;
  bool? isCongresista = false;
  bool? isInvitado = false;
  bool? isDisertante = false;
  bool? isAudioVisual = false;

  bool? isActivado = true;

  Usuario({
    this.id,
    this.nombreCompleto,
    this.email,
    this.senha,
    this.telefono,
    this.institucion,
    this.registroAcademico,
    this.semestre,
    this.seccion,
    this.isPago = false,
    this.fechaPago,
    this.montoPago,
    this.usuarioPago,
    this.uuid,
    this.pais,
    this.fechaRegistro,
    this.recordarPassword,
    this.isAdmin = false,
    this.isFinanciero = false,
    this.isStaff = false,
    this.isCongresista = false,
    this.isExonerado = false,
    this.isActivado = true,
    this.isInvitado = false,
    this.isDisertante = false,
    this.isAudioVisual = false,
  });
  @override
  String get title => nombreCompleto ?? 'No Name';

  factory Usuario.fromJson(Map<String, dynamic> json) =>
      _$UsuarioFromJson(json);
  Map<String, dynamic> toJson() => _$UsuarioToJson(this);

  String getEstado() {
    if (isExonerado == true) {
      return 'Exonerado';
    }
    if (isPago == true) {
      return 'Pagado';
    }
    return 'Pendiente';
  }

  String getTipo() {
    if (isStaff == true) {
      return 'STAFF';
    } else if (isInvitado == true) {
      return 'INVITADO';
    } else if (isDisertante == true) {
      return 'DISERTANTE';
    }
    return 'PARTICIPANTE';
  }

  Usuario copyWith({
    int? id,
    DateTime? fechaRegistro,
    String? nombreCompleto,
    String? email,
    String? senha,
    String? telefono,
    String? institucion,
    String? registroAcademico,
    String? semestre,
    String? seccion,
    String? pais,
    bool? isPago,
    DateTime? fechaPago,
    double? montoPago,
    String? usuarioPago,
    bool? isExonerado,
    String? uuid,
    bool? recordarPassword,
    bool? isAdmin,
    bool? isFinanciero,
    bool? isStaff,
    bool? isCongresista,
    bool? isInvitado,
    bool? isDisertante,
    bool? isActivado,
    bool? isAudioVisual,
  }) {
    return Usuario(
      id: id ?? this.id,
      fechaRegistro: fechaRegistro ?? this.fechaRegistro,
      nombreCompleto: nombreCompleto ?? this.nombreCompleto,
      email: email ?? this.email,
      senha: senha ?? this.senha,
      telefono: telefono ?? this.telefono,
      institucion: institucion ?? this.institucion,
      registroAcademico: registroAcademico ?? this.registroAcademico,
      semestre: semestre ?? this.semestre,
      seccion: seccion ?? this.seccion,
      pais: pais ?? this.pais,
      isPago: isPago ?? this.isPago,
      fechaPago: fechaPago ?? this.fechaPago,
      montoPago: montoPago ?? this.montoPago,
      usuarioPago: usuarioPago ?? this.usuarioPago,
      isExonerado: isExonerado ?? this.isExonerado,
      uuid: uuid ?? this.uuid,
      recordarPassword: recordarPassword ?? this.recordarPassword,
      isAdmin: isAdmin ?? this.isAdmin,
      isFinanciero: isFinanciero ?? this.isFinanciero,
      isStaff: isStaff ?? this.isStaff,
      isCongresista: isCongresista ?? this.isCongresista,
      isInvitado: isInvitado ?? this.isInvitado,
      isDisertante: isDisertante ?? this.isDisertante,
      isActivado: isActivado ?? this.isActivado,
      isAudioVisual: isAudioVisual ?? this.isAudioVisual,
    );
  }
}
